sample-xml/data/src/main/java/co/nimblehq/sample/xml/data/repository/AppPreferencesRepositoryImpl.kt

```diff
@@ -0,0 +1,40 @@
+package co.nimblehq.sample.xml.data.repository
+
+import android.util.Log
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.*
+import co.nimblehq.sample.xml.domain.repository.AppPreferencesRepository
+import kotlinx.coroutines.flow.*
+import java.io.IOException
+import javax.inject.Inject
+
+class AppPreferencesRepositoryImpl @Inject constructor(
```

Comment:
@lydiasama It's good to have unit-test for this Repository class too. I wrote a unit-test class in the new project which applies this new DataStore logic, please feel free to check it out as a reference for your implementation.

🚧  Note that, we might not be able to use MockK to mock & test `updateFirstTimeLaunch`, so I need to prepare a TestDataStore to test it `as an instrument test with Robolectric`.

```
private const val TEST_DATASTORE_NAME: String = "test_datastore"

@ExperimentalCoroutinesApi
@RunWith(RobolectricTestRunner::class)
class AppPreferencesRepositoryTest {

    private val testContext: Context = ApplicationProvider.getApplicationContext()
    private val testCoroutineDispatcher = UnconfinedTestDispatcher()
    private val testCoroutineScope = TestScope(testCoroutineDispatcher + Job())
    private val testDataStore: DataStore<Preferences> = PreferenceDataStoreFactory.create(
        scope = testCoroutineScope,
        produceFile =
        { testContext.preferencesDataStoreFile(TEST_DATASTORE_NAME) }
    )

    private val mockDataStore: DataStore<Preferences> = mockk()
    private lateinit var repository: AppPreferencesRepository

    @Before
    fun setup() {
        repository = AppPreferencesRepositoryImpl(mockDataStore)
    }

    @Test
    fun `When calling getItSupportInfo request successfully, it returns success response`() =
        runTest {
            val mockPreferences = mockk<Preferences>()
            every { mockPreferences[stringPreferencesKey("IT_SUPPORT_PHONE_NUMBER")] } returns "02-134-3556"
            every { mockPreferences[stringPreferencesKey("IT_SUPPORT_PHONE_NUMBER_EXT")] } returns "123"
            every { mockDataStore.data } returns flowOf(mockPreferences)

            repository.getItSupportInfo().test {
                expectMostRecentItem() shouldBe MockUtil.itSupportInfo
            }
        }

    @Test
    fun `When calling getItSupportInfo request failed with IOException, it returns empty ItSupportInfo`() =
        runTest {
            val expected = IOException()
            every { mockDataStore.data } returns flow { throw expected }

            repository.getItSupportInfo().test {
                expectMostRecentItem() shouldBe ItSupportInfo(
                    phoneNumber = "",
                    phoneNumberExtension = ""
                )
            }
        }

    @Test
    fun `When calling getItSupportInfo request failed with other exceptions, it returns wrapped error`() = runTest {
        val expected = Exception()
        every { mockDataStore.data } returns flow { throw expected }

        repository.getItSupportInfo().test {
            awaitError() shouldBe expected
        }
    }

    /**
     * FIXME Can't use MockK to mock DataStore.edit or DataStore.updateData, switch to use a test DataStore.
     */
    @Test
    fun `When calling updateItSupportInfo, it updates ItSupportInfo into DataStore`() =
        runTest {
            repository = AppPreferencesRepositoryImpl(testDataStore)

            val expected = MockUtil.itSupportInfo

            repository.updateItSupportInfo(expected)

            repository.getItSupportInfo().test {
                expectMostRecentItem() shouldBe expected
            }
        }
}
```

Reply:
Please consider adding it to the `sample-compose` project too.

Reply:
@luongvo Resolved in bba361ff